File: src/hangman_game/hangman_game/user_input.py
# hangman_game/user_input.py

import rclpy
from rclpy.node import Node
from hangman_interfaces.srv import CheckLetter
import threading


class UserInput(Node):

    def __init__(self):
        super().__init__("user_input")
        self.cli = self.create_client(CheckLetter, "check_letter")
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("Service not available, waiting...")
        self.req = CheckLetter.Request()
        self.get_logger().info("Press Enter to check the current letter.")
        threading.Thread(target=self.input_thread, daemon=True).start()

    def input_thread(self):
        while True:
            input("Press Enter to input the current letter.")
            self.send_request()

    def send_request(self):
        future = self.cli.call_async(self.req)
        # future.add_done_callback(self.callback_future)

    # def callback_future(self, future):
    #     response = future.result()
    #     self.get_logger().info(f'{response.message}')
    #     self.get_logger().info(f'Word State: {response.updated_word_state}')


def main(args=None):
    rclpy.init(args=args)
    user_input = UserInput()
    rclpy.spin(user_input)
    user_input.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/hangman_game_cpp/src/user_input.cpp
#include <chrono>
#include <memory>
#include <string>
#include <thread>
#include <iostream>

#include "rclcpp/rclcpp.hpp"
#include "hangman_interfaces/srv/check_letter.hpp"

using namespace std::chrono_literals;
using CheckLetter = hangman_interfaces::srv::CheckLetter;

class UserInput : public rclcpp::Node
{
public:
  UserInput()
  : Node("user_input")
  {
    client_ = this->create_client<CheckLetter>("check_letter");

    while (!client_->wait_for_service(1s)) {
      RCLCPP_INFO(this->get_logger(), "Service not available, waiting...");
    }

    RCLCPP_INFO(this->get_logger(), "Press Enter to check the current letter.");
    std::thread(&UserInput::input_loop, this).detach();
  }

private:
  rclcpp::Client<CheckLetter>::SharedPtr client_;

  void input_loop()
  {
    while (rclcpp::ok()) {
      std::string dummy;
      std::getline(std::cin, dummy);  // 엔터 대기

      auto request = std::make_shared<CheckLetter::Request>();
      auto result_future = client_->async_send_request(request,
        std::bind(&UserInput::handle_response, this, std::placeholders::_1));
    }
  }

  void handle_response(rclcpp::Client<CheckLetter>::SharedFuture future)
  {
    auto response = future.get();
    RCLCPP_INFO(this->get_logger(), "Response message: %s", response->message.c_str());
    RCLCPP_INFO(this->get_logger(), "Word State: %s", response->updated_word_state.c_str());
  }
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<UserInput>();
  rclcpp::spin(node);  // input_thread는 detach 상태로 계속 대기
  rclcpp::shutdown();
  return 0;
}